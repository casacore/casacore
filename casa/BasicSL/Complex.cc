//# Complex.cc: Implement Complex, DComplex
//# Copyright (C) 2000,2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


//# Includes
#include <casacore/casa/BasicSL/Complex.h>
#include <casacore/casa/BasicMath/Math.h>
#include <casacore/casa/BasicSL/Constants.h>

namespace casa { //# NAMESPACE CASA - BEGIN

// Math functions
/// Should be in stl

#if defined(NEEDS_LOG10_COMPLEX)
DComplex log10(const DComplex &val)
{
  return DComplex(std::log(val)*C::log10e);
}

/// Should be in stl
Complex log10(const Complex &val)
{
  // Need to make log10e a Float for it to compile
  // with picky compilers
  return Complex(std::log(val)*Float(C::log10e));
}
#endif

// Near functions
// Note: max() cannot be used from Math.h until it is derived from <math>
// Note: abs() not defined in SGI
//
Bool near(const Complex &val1, const Complex &val2, Double tol)
{
  if (tol <= 0) return val1 == val2;
  if (val1 == val2) return True;
  if (near(val1.real(), val2.real(), tol) &&
      near(val1.imag(), val2.imag(), tol)) return True;
  Float aval1(std::abs(val1)), aval2(std::abs(val2));
  if (aval1 == 0) return aval2 <= (1+tol)*FLT_MIN;
  else if (aval2 == 0) return aval1 <= (1+tol)*FLT_MIN;
  DComplex dval(val1);
  dval -= DComplex(val2);
  return std::abs(dval) <= tol * (aval1 < aval2 ? aval2 : aval1);
}

Bool near(const DComplex &val1, const DComplex &val2, Double tol)
{
  if (tol <= 0) return val1 == val2;
  if (val1 == val2) return True;
  if (std::abs(val1) == 0) return std::abs(val2) <= (1+tol)*DBL_MIN;
  else if (std::abs(val2) == 0) return std::abs(val1) <= (1+tol)*DBL_MIN;
  Double aval1(std::abs(val1)), aval2(std::abs(val2));
  return std::abs(val1-val2) <= tol * (aval1 < aval2 ? aval2 : aval1);
}

Bool nearAbs(const Complex &val1, const Complex &val2, Double tol)
{
  return std::abs(val2 - val1) <= tol;
}
Bool nearAbs(const DComplex &val1, const DComplex &val2, Double tol)
{
  return std::abs(val2 - val1) <= tol;
}

// NaN functions

Bool isNaN(const Complex &val)
{
  return isNaN(val.real()) || isNaN(val.imag());
}
Bool isNaN(const DComplex &val)
{
  return isNaN(val.real()) || isNaN(val.imag());
}
void setNaN(Complex &val)
{
  Float x; setNaN(x);
  Float y; setNaN(y);
  val = Complex(x, y);
}
void setNaN(DComplex &val)
{
  Double x; setNaN(x);
  Double y; setNaN(y);
  val = DComplex(x, y);
}

// Inf functions

Bool isInf(const Complex &val)
{
  return isInf(val.real()) || isInf(val.imag());
}
Bool isInf(const DComplex &val)
{
  return isInf(val.real()) || isInf(val.imag());
}
void setInf(Complex &val)
{
  Float x; setInf(x);
  Float y; setInf(y);
  val = Complex(x, y);
}
void setInf(DComplex &val)
{
  Double x; setInf(x);
  Double y; setInf(y);
  val = DComplex(x, y);
}


// Finite functions

Bool isFinite(const Complex &val)
{
  return isFinite(val.real()) || isFinite(val.imag());
}
Bool isFinite(const DComplex &val)
{
  return isFinite(val.real()) || isFinite(val.imag());
}

// fmod functions

DComplex fmod(const DComplex &in, const DComplex &f) {
  return DComplex(std::fmod(real(in), real(f)), imag(in)); }
Complex fmod(const Complex &in, const Complex &f) {
  return Complex(std::fmod(real(in), real(f)), imag(in)); }

// Inverse trigonometry (see Abromowitz)
DComplex atan(const DComplex &in) {
  const Double n = norm(in);
  return DComplex(0.5*std::atan(2.0*real(in)/(1.0-n)),
		  0.25*std::log((1.0+n+2*imag(in))/(1.0+n-2*imag(in))));
}
Complex atan(const Complex &in) {
  const Float n = norm(in);
  return Complex(0.5*std::atan(2.0*real(in)/(1.0-n)),
		 0.25*std::log((1.0+n+2*imag(in))/(1.0+n-2*imag(in))));
}
DComplex asin(const DComplex &in) {
  const Double n = norm(in);
  Double a = 0.5*std::sqrt(1.0+n+2*real(in));
  const Double c = 0.5*std::sqrt(1.0+n-2*real(in));
  const Double b = a-c;
  a += c;
  return DComplex(std::asin(b), std::log(a+std::sqrt(a*a-1.0)));
}
Complex asin(const Complex &in) {
  const Float n = norm(in);
  Float a = 0.5*std::sqrt(1.0+n+2*real(in));
  const Float c = 0.5*sqrt(1.0+n-2*real(in));
  const Float b = a-c;
  a += c;
  return Complex(std::asin(b), std::log(a+std::sqrt(a*a-1.0)));
}
DComplex acos(const DComplex &in) {
  const Double n = norm(in);
  Double a = 0.5*std::sqrt(1.0+n+2*real(in));
  const Double c = 0.5*std::sqrt(1.0+n-2*real(in));
  const Double b = a-c;
  a += c;
  return DComplex(std::acos(b), -std::log(a+std::sqrt(a*a-1.0)));
}
Complex acos(const Complex &in) {
  const Float n = norm(in);
  Float a = 0.5*std::sqrt(1.0+n+2*real(in));
  const Float c = 0.5*std::sqrt(1.0+n-2*real(in));
  const Float b = a-c;
  a += c;
  return Complex(std::acos(b), -std::log(a+std::sqrt(a*a-1.0)));
}
DComplex atan2(const DComplex &in, const DComplex &t2) {
  if (norm(t2) == 0) return DComplex(C::pi_2);
  const DComplex z = atan(in/t2);
  if (real(t2) > 0) return z;
  return (z + DComplex(C::pi));
}
Complex atan2(const Complex &in, const Complex &t2) {
  if (norm(t2) == 0) return Complex(C::pi_2);
  const Complex z = atan(in/t2);
  if (real(t2) > 0) return z;
  return (z + Complex(C::pi));
}
/// Temporary solutions only
DComplex erf(const DComplex &in) {
  return ::erf(in.real());
}
Complex erf(const Complex &in) {
  return ::erf(in.real());
}
DComplex erfc(const DComplex &in) {
  return ::erfc(in.real());
}
Complex erfc(const Complex &in) {
  return ::erfc(in.real());
}

} //# NAMESPACE CASA - END
