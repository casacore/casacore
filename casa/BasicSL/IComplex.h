//# IComplex.h: Integer complex number
//# Copyright (C) 2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


#ifndef CASA_ICOMPLEX_H
#define CASA_ICOMPLEX_H

//# Includes

#include <casacore/casa/aips.h>
#include <casacore/casa/iosfwd.h>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

// <summary>
// Integer complex numbers.
// </summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>

// <synopsis>
// The class <src>IComplex</src> is only a container for FITS usage.
// </synopsis>

class IComplex {
public:
  //# Constructors
  // From one or two ints (note for gnu use)
  // <group>
  IComplex() : re(0), im(0) {;};
  IComplex(Int r) : re(r), im(0) {;};
  IComplex(Int r, Int i) : re(r), im(i) {;};
  // </group>

  //# Member functions
  // For use in FITS classes only
  // <group>
  Int real() const { return re; };
  Int imag() const { return im; };
  // </group>

private:
  Int re;
  Int im;
};


// Show on ostream.
ostream &operator<< (ostream &os, const IComplex&);



} //# NAMESPACE CASACORE - END

#endif
