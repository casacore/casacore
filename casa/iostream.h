//# iostream.h: Interim solution for standard/nonstandard system iostream
//# Copyright (C) 2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_IOSTREAM_H
#define CASA_IOSTREAM_H

//# Define the C standard C++ include file. 
//# This is an interim solution to cater for the SGI non-existence of
//# them (e.g. <cstring>)

//# Make sure any special macros are set
#include <casacore/casa/aips.h>
#include <iostream>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

using std::istream;
using std::ostream;
using std::iostream;
using std::cin;
using std::cout;
using std::cerr;
using std::endl;
using std::flush;
using std::ws;

using std::ios;
using std::streampos;
using std::streamoff;
using std::streamsize;
using std::dec;
using std::hex;
using std::oct;

using std::internal;
using std::left;
using std::right;
using std::fixed;
using std::scientific;

using std::boolalpha;
using std::noboolalpha;
using std::showbase;
using std::noshowbase;
using std::showpoint;
using std::noshowpoint;
using std::showpos;
using std::noshowpos;
using std::skipws;
using std::noskipws;
using std::uppercase;
using std::nouppercase;

} //# NAMESPACE CASACORE - END

#endif
