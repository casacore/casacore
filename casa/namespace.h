#ifndef NAMESPACE_H
#define NAMESPACE_H 
using namespace casa;
#endif
