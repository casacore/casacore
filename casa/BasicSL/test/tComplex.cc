//# tComplex.cc: This program tests the Complex class
//# Copyright (C) 1993,1994,1995,1996,1999,2000,2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes

#include <casa/BasicSL/Complex.h>
#include <casa/BasicSL/IComplex.h>

#include <casa/Utilities/Assert.h>
#include <casa/stdio.h>
#include <casa/iostream.h>
#include <casa/fstream.h>
#include <unistd.h>

#include <casa/namespace.h>
int main() {

  Complex f1(23.9,1.8), f2(9.2,8.2), f3(2.7,1.8), fo(237.561,0.9312), fi;
  IComplex i1(5,2), i3(Int(f1.real()),Int(f1.imag()));
  DComplex d1, d2(f1.real(),f1.imag()), d3(0.921,7.812);
  fstream fio("tComplex_tmp.data", ios::out | ios::trunc);

  cout << "Initial value for complex: " << d1 << endl;
  cout << d1 << " :" << endl;
  cout << "    real part:      " << d1.real() << " == " << real(d1) << endl;
  cout << "    imaginary part: " << d1.imag() << " == " << imag(d1) << endl;
  cout << f1 << " == " << d2 << " ~= " << i3 << endl;
  cout << "-" << d3 << " == " << -d3 << endl;
  cout << "conj(" << d3 << ") == " << conj(d3) << endl;
  cout << "norm(" << d3 << ") == " << norm(d3) << endl;
  cout << "arg(" << d3 << ") == " << arg(d3) << endl;
  d1 = DComplex(18.9, -2.31);
  cout << "fabs(" << d1 << ") == " << fabs(d1) << endl;
  fio << fo << endl;
  fio.close();
  fio.open("tComplex_tmp.data", ios::in);
  fio >> fi;
  fio.close();
  unlink("tComplex_tmp.data");
  cout << "out: " << fo << "           in: " << fi << endl;

  
  d1 = DComplex(i1.real(), i1.imag());
  cout << d1 << " == " << i1 << endl;

  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  cout << f1 << " + " << f2 << " = " << f1 + f2 << endl;
  cout << f1 << " * " << f2 << " = " << f1 * f2 << endl;
  cout << f1 << " - " << f2 << " = " << f1 - f2 << endl;
  cout << f1 << " / " << f2 << " = " << f1 / f2 << endl;
  cout << endl;

  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  cout << f1 << " + " << 7.561 << " = " << f1 + 7.561f << endl;
  cout << f1 << " * " << 7.561 << " = " << f1 * 7.561 << endl;
  cout << f1 << " - " << 7.561 << " = " << f1 - 7.561f << endl;
  cout << f1 << " / " << 7.561 << " = " << f1 / 7.561 << endl;
  cout << endl;

  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  cout << f1 << " + " << 13 << " = " << f1 + 13.0f << endl;
  cout << f1 << " * " << "13.0f" << " = " << f1 * 13.0f << endl;
  cout << f1 << " * " << "13" << " = " << f1 * 13 << endl;
  cout << f1 << " * " << "13." << " = " << f1 * 13. << endl;
  cout << "13.0f" << " * " << f1 << " = " << 13.0f * f1 << endl;
  cout << "13" << " * " << f1 << " = " << 13 * f1 << endl;
  cout << "13." << " * " << f1 << " = " << 13. * f1 << endl;
  cout << f1 << " - " << 13 << " = " << f1 - 13.0f << endl;
  cout << f1 << " / " << "13.0f" << " = " << f1 / 13.0f << endl;
  cout << f1 << " / " << "13" << " = " << f1 / 13 << endl;
  cout << f1 << " / " << "13." << " = " << f1 / 13. << endl;
  cout << "13.0f" << " / " << f1 << " = " << 13.0f / f1 << endl;
  cout << "13" << " / " << f1 << " = " << 13 / f1 << endl;
  cout << "13." << " / " << f1 << " = " << 13. / f1 << endl;
  cout << endl;

  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  cout << f1 << " +=  4 -> "; cout << (f1 += 4) << endl;
  cout << f1 << " *=  4 -> "; cout << (f1 *= 4) << endl;
  cout << f1 << " -=  4 -> "; cout << (f1 -= 4) << endl;
  cout << f1 << " /=  4 -> "; cout << (f1 /= 4) << endl;
  cout << endl;

  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  cout << f1 << " +=  0.896 -> "; cout << (f1 += 0.896) << endl;
  cout << f1 << " *=  0.896 -> "; cout << (f1 *= 0.896) << endl;
  cout << f1 << " -=  0.896 -> "; cout << (f1 -= 0.896) << endl;
  cout << f1 << " /=  0.896 -> "; cout << (f1 /= 0.896) << endl;
  cout << endl;

  char simChar = 2;
  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  cout << f1 << " +=  simChar -> "; cout << (f1 += simChar) << endl;
  cout << f1 << " *=  simChar -> "; cout << (f1 *= simChar) << endl;
  cout << f1 << " -=  simChar -> "; cout << (f1 -= simChar) << endl;
  cout << f1 << " /=  simChar -> "; cout << (f1 /= simChar) << endl;
  cout << endl;

  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  cout << f1 << " +=  " << f3 << " -> "; cout << (f1 += f3) << endl;
  cout << f1 << " *=  " << f3 << " -> "; cout << (f1 *= f3) << endl;
  cout << f1 << " -=  " << f3 << " -> "; cout << (f1 -= f3) << endl;
  cout << f1 << " /=  " << f3 << " -> "; cout << (f1 /= f3) << endl;
  cout << endl;

  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  Complex d3c = Complex(d3.real(), d3.imag());
  cout << f1 << " +=  " << d3 << " -> "; cout << (f1 += d3c) << endl;
  cout << f1 << " *=  " << d3 << " -> "; cout << (f1 *= d3c) << endl;
  cout << f1 << " -=  " << d3 << " -> "; cout << (f1 -= d3c) << endl;
  cout << f1 << " /=  " << d3 << " -> "; cout << (f1 /= d3c) << endl;
  cout << endl;

  i1 = IComplex(36, 14);

  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  // reset f1 to nominal values at it will have been corrupted (due to
  // roundoff) by all the earlier calculations.
  f1 = Complex(33.417, 13.412);
  cout << "sin(" << f1 << ") = " << sin(f1) << endl;
  cout << "cos(" << f1 << ") = " << cos(f1) << endl;
  cout << "sinh(" << f1 << ") = " << sinh(f1) << endl;
  cout << "cosh(" << f1 << ") = " << cosh(f1) << endl;
  cout << endl;

  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  cout << "log(" << f1 << ") = " << log(f1) << endl;
  cout << "log10(" << f1 << ") = " << log10(f1) << endl;
  cout << "exp(" << f1 << ") = " << exp(f1) << endl;
  cout << endl;

  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  if (! near (pow(f1,3.0f), Complex(19283.3,42518.8))) {
    cout << "pow(" << f1 << ",3) = " << pow(f1,3.0f) << endl;
  }
  if (! near (pow(f1,-8.0f), Complex(-3.52463e-13,-3.11743e-14))) {
    cout << "pow(" << f1 << ",-8) = " << pow(f1,-8.0f) << endl;
  }
  if (! near (pow(f1,0.214f), Complex(2.14595,0.175667))) {
    cout << "pow(" << f1 << ",0.214) = " << pow(f1,0.214f) << endl;
  }
  cout << "pow(" << f1 << "," << i1 << " = "
       << pow(f1,Complex(i1.real(), i1.imag())) << endl;
  if (! near (pow(f1,f2), Complex(8.06899e+11,9.07712e+12))) {
    cout << "pow(" << f1 << "," << f2 << " = " << pow(f1,f2) << endl;
  }
  cout << endl;

  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  cout << "sqrt(" << f1 << ") = " << sqrt(f1) << endl;
  cout << endl;

  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  cout << f1 << " == " << f2 << " -> " << (f1 == f2) << endl;
  cout << f1 << " != " << f2 << " -> " << (f1 != f2) << endl;
  cout << f1 << " >= " << f2 << " -> " << (f1 >= f2) << endl;
  cout << f1 << " > " << f2 << " -> " << (f1 > f2) << endl;
  cout << f1 << " <= " << f2 << " -> " << (f1 <= f2) << endl;
  cout << f1 << " < " << f2 << " -> " << (f1 < f2) << endl;
  cout << endl;

  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  cout << f2 << " == " << f1 << " -> " << (f2 == f1) << endl;
  cout << f2 << " != " << f1 << " -> " << (f2 != f1) << endl;
  cout << f2 << " >= " << f1 << " -> " << (f2 >= f1) << endl;
  cout << f2 << " > " << f1 << " -> " << (f2 > f1) << endl;
  cout << f2 << " <= " << f1 << " -> " << (f2 <= f1) << endl;
  cout << f2 << " < " << f1 << " -> " << (f2 < f1) << endl;
  cout << endl;

  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;
  cout << f1 << " == " << f1 << " -> " << (f1 == f1) << endl;
  cout << f1 << " != " << f1 << " -> " << (f1 != f1) << endl;
  cout << f1 << " >= " << f1 << " -> " << (f1 >= f1) << endl;
  cout << f1 << " > " << f1 << " -> " << (f1 > f1) << endl;
  cout << f1 << " <= " << f1 << " -> " << (f1 <= f1) << endl;
  cout << f1 << " < " << f1 << " -> " << (f1 < f1) << endl;
  cout << endl;

  f1 = Complex(33.0, 6.0);

  f1 = 33.021;

  f1 = 0;
  cout << "- - - - - - - - - - - - - - - - - - - -" << endl;

  AlwaysAssertExit(near(Complex(0,10000), Complex(0,10001), 1.01e-4));
  AlwaysAssertExit(!near(Complex(0,10000), Complex(0,10001), 0.99e-4));
  AlwaysAssertExit(!near(Complex(10000,0), Complex(0,10001), 1.01e-4));
  AlwaysAssertExit(nearAbs(Complex(0,10000), Complex(0,10001), 1.01));
  AlwaysAssertExit(!nearAbs(Complex(0,10000), Complex(0,10001), 0.99));
  AlwaysAssertExit(!nearAbs(Complex(10000,0), Complex(0,10001), 1.01));

  AlwaysAssertExit(allNear(Complex(0,10000), Complex(0,10001), 1.01e-4));
  AlwaysAssertExit(!allNear(Complex(0,10000), Complex(0,10001), 0.99e-4));
  AlwaysAssertExit(!allNear(Complex(10000,0), Complex(0,10001), 1.01e-4));
  AlwaysAssertExit(allNearAbs(Complex(0,10000), Complex(0,10001), 1.01));
  AlwaysAssertExit(!allNearAbs(Complex(0,10000), Complex(0,10001), 0.99));
  AlwaysAssertExit(!allNearAbs(Complex(10000,0), Complex(0,10001), 1.01));

  AlwaysAssertExit(near(DComplex(0,10000), DComplex(0,10001), 1.01e-4));
  AlwaysAssertExit(!near(DComplex(0,10000), DComplex(0,10001), 0.99e-4));
  AlwaysAssertExit(!near(DComplex(10000,0), DComplex(0,10001), 1.01e-4));
  AlwaysAssertExit(nearAbs(DComplex(0,10000), DComplex(0,10001), 1.01));
  AlwaysAssertExit(!nearAbs(DComplex(0,10000), DComplex(0,10001), 0.99));
  AlwaysAssertExit(!nearAbs(DComplex(10000,0), DComplex(0,10001), 1.01));

  AlwaysAssertExit(allNear(DComplex(0,10000), DComplex(0,10001), 1.01e-4));
  AlwaysAssertExit(!allNear(DComplex(0,10000), DComplex(0,10001), 0.99e-4));
  AlwaysAssertExit(!allNear(DComplex(10000,0), DComplex(0,10001), 1.01e-4));
  AlwaysAssertExit(allNearAbs(DComplex(0,10000), DComplex(0,10001), 1.01));
  AlwaysAssertExit(!allNearAbs(DComplex(0,10000), DComplex(0,10001), 0.99));
  AlwaysAssertExit(!allNearAbs(DComplex(10000,0), DComplex(0,10001), 1.01));

  Complex c1; DComplex c2;
  setNaN(c1); setNaN(c2);
  AlwaysAssertExit(isNaN(c1) && isNaN(c2));
  c1 = Complex(0.0, c1.imag()); c2 = DComplex(c2.real(), 0.0);
  AlwaysAssertExit(isNaN(c1) && isNaN(c2));
  c1 = Complex(0.0); c2 = DComplex(0.0);
  AlwaysAssertExit((!isNaN(c1)) && (!isNaN(c2)));

  {
      // Test min/max
      Complex c1(0,1), c2(2,0);
      Complex c3 = min(c1,c2);
      AlwaysAssertExit(near(c1,c3));
      Complex c4 = max(c1,c2);
      AlwaysAssertExit(near(c2,c4));
  }

  return(0);
}
