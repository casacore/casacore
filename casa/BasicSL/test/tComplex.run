#!/bin/sh

# Change Infinity, infinity or inf to Inf and remove all leading/trailing whitespace 
$casa_checktool ./tComplex | sed -e 's/nfinity/nf/g;s/inf/Inf/g;s/ *Inf */Inf/g'
