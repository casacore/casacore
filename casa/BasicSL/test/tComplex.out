Initial value for complex: (0,0)
(0,0) :
    real part:      0 == 0
    imaginary part: 0 == 0
(23.9,1.8) == (23.9,1.8) ~= (23,1)
-(0.921,7.812) == (-0.921,-7.812)
conj((0.921,7.812)) == (0.921,-7.812)
norm((0.921,7.812)) == 61.8756
arg((0.921,7.812)) == 1.45344
fabs((18.9,-2.31)) == 19.0406
out: (237.561,0.9312)           in: (237.561,0.9312)
(5,2) == (5,2)
- - - - - - - - - - - - - - - - - - - -
(23.9,1.8) + (9.2,8.2) = (33.1,10)
(23.9,1.8) * (9.2,8.2) = (205.12,212.54)
(23.9,1.8) - (9.2,8.2) = (14.7,-6.4)
(23.9,1.8) / (9.2,8.2) = (1.5449,-1.18133)

- - - - - - - - - - - - - - - - - - - -
(23.9,1.8) + 7.561 = (31.461,1.8)
(23.9,1.8) * 7.561 = (180.708,13.6098)
(23.9,1.8) - 7.561 = (16.339,1.8)
(23.9,1.8) / 7.561 = (3.16096,0.238064)

- - - - - - - - - - - - - - - - - - - -
(23.9,1.8) + 13 = (36.9,1.8)
(23.9,1.8) * 13 = (310.7,23.4)
(23.9,1.8) - 13 = (10.9,1.8)
(23.9,1.8) / 13 = (1.83846,0.138462)

- - - - - - - - - - - - - - - - - - - -
(23.9,1.8) +=  4 -> (27.9,1.8)
(27.9,1.8) *=  4 -> (111.6,7.2)
(111.6,7.2) -=  4 -> (107.6,7.2)
(107.6,7.2) /=  4 -> (26.9,1.8)

- - - - - - - - - - - - - - - - - - - -
(26.9,1.8) +=  0.896 -> (27.796,1.8)
(27.796,1.8) *=  0.896 -> (24.9052,1.6128)
(24.9052,1.6128) -=  0.896 -> (24.0092,1.6128)
(24.0092,1.6128) /=  0.896 -> (26.796,1.8)

- - - - - - - - - - - - - - - - - - - -
(26.796,1.8) +=  simChar -> (28.796,1.8)
(28.796,1.8) *=  simChar -> (57.592,3.6)
(57.592,3.6) -=  simChar -> (55.592,3.6)
(55.592,3.6) /=  simChar -> (27.796,1.8)

- - - - - - - - - - - - - - - - - - - -
(27.796,1.8) +=  (2.7,1.8) -> (30.496,3.6)
(30.496,3.6) *=  (2.7,1.8) -> (75.8592,64.6128)
(75.8592,64.6128) -=  (2.7,1.8) -> (73.1592,62.8128)
(73.1592,62.8128) /=  (2.7,1.8) -> (29.496,3.6)

- - - - - - - - - - - - - - - - - - - -
(29.496,3.6) +=  (0.921,7.812) -> (30.417,11.412)
(30.417,11.412) *=  (0.921,7.812) -> (-61.1365,248.128)
(-61.1365,248.128) -=  (0.921,7.812) -> (-62.0575,240.316)
(-62.0575,240.316) /=  (0.921,7.812) -> (29.417,11.412)

- - - - - - - - - - - - - - - - - - - -
sin((33.417,13.412)) = (303543,-139313)
cos((33.417,13.412)) = (-139313,-303543)
sinh((33.417,13.412)) = (1.08012e+14,1.21875e+14)
cosh((33.417,13.412)) = (1.08012e+14,1.21875e+14)

- - - - - - - - - - - - - - - - - - - -
log((33.417,13.412)) = (3.58374,0.381672)
log10((33.417,13.412)) = (1.5564,0.165758)
exp((33.417,13.412)) = (2.16024e+14,2.43751e+14)

- - - - - - - - - - - - - - - - - - - -
pow((33.417,13.412),(36,14) = (Inf,Inf)

- - - - - - - - - - - - - - - - - - - -
sqrt((33.417,13.412)) = (5.89173,1.13821)

- - - - - - - - - - - - - - - - - - - -
(33.417,13.412) == (9.2,8.2) -> 0
(33.417,13.412) != (9.2,8.2) -> 1
(33.417,13.412) >= (9.2,8.2) -> 1
(33.417,13.412) > (9.2,8.2) -> 1
(33.417,13.412) <= (9.2,8.2) -> 0
(33.417,13.412) < (9.2,8.2) -> 0

- - - - - - - - - - - - - - - - - - - -
(9.2,8.2) == (33.417,13.412) -> 0
(9.2,8.2) != (33.417,13.412) -> 1
(9.2,8.2) >= (33.417,13.412) -> 0
(9.2,8.2) > (33.417,13.412) -> 0
(9.2,8.2) <= (33.417,13.412) -> 1
(9.2,8.2) < (33.417,13.412) -> 1

- - - - - - - - - - - - - - - - - - - -
(33.417,13.412) == (33.417,13.412) -> 1
(33.417,13.412) != (33.417,13.412) -> 0
(33.417,13.412) >= (33.417,13.412) -> 1
(33.417,13.412) > (33.417,13.412) -> 0
(33.417,13.412) <= (33.417,13.412) -> 1
(33.417,13.412) < (33.417,13.412) -> 0

- - - - - - - - - - - - - - - - - - - -
